### DISCUSSION TITLE
How to send File Object from client component to server actions in Next.js 14?

### DISCUSSION THREAD

**Discussion by krishmarsonia** [vercel/next.js#67683] [Help]
### Summary

I have been trying to upload the image/video to AWS S3 storage via preSignedURL. After I get the URL from the backend I am sending the URL and File component to the server actions to perform put request on AWS S3. Now I know that I can't send the File Object directly from client to server components. So I First tried to JSON.stringify the File Object, but it was of no use as the JSON.stringify method was returning empty object. So, now I tried to send the FormData to the server component instead of File object, hopefully it was supported but then the new problem aroused, which was when I tried to get the file from the FormData, but I came up with a new error which was "Data after transformation must be a string, an ArrayBuffer, a Buffer, or a Stream". So, now i am trying to send the data in ArrayBuffer and then applying JSON.stringify to it, and now, again it is converting ArrayBuffer into the empty object. What can I do now? please help.

**page.tsx (click handler function)**
```
const clickHandler = async () => {
    const imageName = selected?.name;
    const imageType = selected?.type;

    const result = await createPreSignedObject({
      variables: { name: imageName, contentType: imageType },
    });
    const resultedResult = result.data.createPreSignedObject;
    console.log(resultedResult); //preSignedURL
    if (resultedResult && typeof resultedResult === "string") {
      if (selected) {
        //const formData = new FormData();
        //formData.append("file", selected);
        const fileToArrayBuffer = (file: File) => {
          return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.onload = () => resolve(reader.result);
            reader.onerror = () => reject(reader.error);
            reader.readAsArrayBuffer(file);
          });
        };
        const fileBufferResult = await fileToArrayBuffer(selected);
        console.log(fileBufferResult);
        console.log(JSON.stringify(fileBufferResult)); //giving empty object
        await postObjectToS3(resultedResult, JSON.stringify(fileToArrayBuffer(selected)));
      }
    } else {
      console.log("no result found");
    }
  };
```

**preSignedObjects.ts**
```
"use server"
export const postObjectToS3 = async (url: string, selectedFile: string) => {
  const newResult = JSON.parse(selectedFile);
  console.log(newResult);
  const result = await axios.put(url, newResult, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
};
```
Basically, whatever I have tried JSON.stringify has converted it to empty object only.

**Comment by icyJoseph** [comment]
You should be able to pass `FormData` to `postObjectToS3`. See: https://stackblitz.com/edit/nextjs-qre3a7?file=app%2Factions.ts,app%2Fpage.tsx,app%2FButton.tsx

```ts
'use server';

export const postObjectToS3 = async (url: string, data: FormData) => {
  console.log('file', data.get('imageAsset'), { url });
};

```

Also, axios can recognize, that the data is of formData type, and it will generate the appropriate headers for it. Keep in mind that that, the `multipart/form-data` needs a boundary. Axios will generate it for you, and so will, fetch, why not just use fetch?
